### PR TITLE
[FEAT]: 1inch token swaps

### PR DESCRIPTION
**Pull request by Gamboster**
This PR requires these bitcore changes: 
- https://github.com/bitpay/bitcore/pull/3195
- https://github.com/bitpay/bitcore/pull/3236
- https://github.com/bitpay/bitcore/pull/3237

Next steps
- Send Max feature [DONE]
- Send to any ERC20 token. [DONE]
- Merge with custom ERC20 tokens feature [DONE]
- Merge with current "Exchange Crypto" feature [DONE]

BitPay_JP

### CODE REVIEW

**Review by Chutancdn — APPROVED** [review]

**Review by Chutancdn — APPROVED** [review]

**Review by Chutancdn — DISMISSED** [review]

File: src/components/wallet-selector/wallet-selector.ts [hunk 1]
```diff
@@ -41,4 +47,8 @@ export class WalletSelectorComponent extends ActionSheetParent {
       this.dismiss(optionClicked);
     }
   }
+
+  public readMore() {
+    this.logger.debug('Read more clicked'); // TODO: add a link or remove this read more
```
Comment (rastajpa) [inline src/components/wallet-selector/wallet-selector.ts:52]: update link or hide button to read more option for now...

File: src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+  NavParams,
+  ViewController
+} from 'ionic-angular';
+import * as _ from 'lodash';
+import env from '../../environments';
+
+// Providers
+import {
+  ActionSheetProvider,
+  // BuyCryptoProvider,
```
Comment (rastajpa) [inline src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts:16]: remove unused line

File: src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+  private currentTokenListPage: number;
+  private TOKEN_SHOW_LIMIT: number;
+  public tokenListShowMore: boolean;
+
+  public tokenSearchResults;
+  public selectedToken;
+
+  constructor(
+    private actionSheetProvider: ActionSheetProvider,
+    // private buyCryptoProvider: BuyCryptoProvider,
```
Comment (rastajpa) [inline src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts:61]: same here

File: src/pages/exchange-crypto/exchange-crypto.ts [hunk 4]
```diff
@@ -66,217 +103,508 @@ export class ExchangeCryptoPage {
...
+      switch (promise.exchange) {
+        case '1inch':
+          if (!promise.data || !promise.data.tokens) {
+            this.logger.error('1Inch getCurrencies Error');
+            // this.showErrorAndBack(
+            //   null,
+            //   this.translate.instant(
+            //     '1Inch is not available at this moment. Please, try again later.'
+            //   )
+            // );
```
Comment (rastajpa) [inline src/pages/exchange-crypto/exchange-crypto.ts:230]: what about this error?

File: src/pages/exchange-crypto/exchange-crypto.ts [hunk 5]
```diff
@@ -66,217 +103,508 @@ export class ExchangeCryptoPage {
...
+          if (promise.data.error) {
+            this.logger.error(
+              'Changelly getCurrencies Error: ' + promise.data.error.message
+            );
+            // this.showErrorAndBack(
+            //   null,
+            //   this.translate.instant(
+            //     'Changelly is not available at this moment. Please, try again later.'
+            //   )
+            // );
```
Comment (rastajpa) [inline src/pages/exchange-crypto/exchange-crypto.ts:278]: and this one

File: src/pages/exchange-crypto/exchange-crypto.ts [hunk 6]
```diff
@@ -66,217 +103,508 @@ export class ExchangeCryptoPage {
...
+          token.symbol.toLowerCase() != this.fromWalletSelected.coin &&
+          token.symbol.toLowerCase() != 'eth'
+        );
+      }
+    );
+
+    let modal = this.modalCtrl.create(
+      CoinAndWalletSelectorPage,
+      {
+        title: this.toWalletSelectorTitle,
```
Comment (rastajpa) [inline src/pages/exchange-crypto/exchange-crypto.ts:496]: replace `title` for `walletSelectorTitle`

File: src/pages/send/amount/amount.ts [hunk 7]
```diff
@@ -95,6 +95,7 @@ export class AmountPage {
   private alternativeCurrency;
   public fromBuyCrypto: boolean;
   public fromExchangeCrypto: boolean;
+  public fromTokenSwap: boolean;
```
Comment (rastajpa) [inline src/pages/send/amount/amount.ts:98]: this flag looks is unused

**Review by rastajpa — COMMENTED** [review]
excellent work Gabriel! The feature looks so good! I just sent a few small comments of some things that I saw so far ... I will continue to test the feature and review the code

File: src/components/wallet-selector/wallet-selector.ts (hunk 1 above)
Comment (Gamboster) [reply src/components/wallet-selector/wallet-selector.ts]: Done

File: src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts (hunk 2 above)
Comment (Gamboster) [reply src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts]: Done

File: src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts (hunk 3 above)
Comment (Gamboster) [reply src/pages/coin-and-wallet-selector/coin-and-wallet-selector.ts]: Done

File: src/pages/exchange-crypto/exchange-crypto.ts (hunk 4 above)
Comment (Gamboster) [reply src/pages/exchange-crypto/exchange-crypto.ts]: This error handle is no longer necessary. I removed it

File: src/pages/exchange-crypto/exchange-crypto.ts (hunk 5 above)
Comment (Gamboster) [reply src/pages/exchange-crypto/exchange-crypto.ts]: This error handle is no longer necessary. I removed it

File: src/pages/exchange-crypto/exchange-crypto.ts (hunk 6 above)
Comment (Gamboster) [reply src/pages/exchange-crypto/exchange-crypto.ts]: Done

File: src/pages/send/amount/amount.ts (hunk 7 above)
Comment (Gamboster) [reply src/pages/send/amount/amount.ts]: Done. Thanks!

**Review by cmgustavo — APPROVED** [review]
ACK! Great work 🚀

**Review by cmgustavo — APPROVED** [review]

**Pull request merged**